### PR TITLE
WIP: HTTP Transport

### PR DESCRIPTION
**Pull request by Skeen**
### What does this PR do?
* Introduces `salt/transport/abstract,py`, which provides abstract implementations of the channels by refactoring out sharable behavior from the `tcp` transport (`salt/transport/tcp.py`).
This is still work-in-progress, and I'd like to refactor it to a mixin-based architecture, where the channel flows through mixins / layers / decorators towards the transport itself, such that the transport layer itself should merely provide an `rx` and a `tx` function for each direction of the data-flow.
* Introduces `salt/transport/http.py`, which utilizes the abstract implementations from above, to implement a `http` transport using a message queue and long-polling.

The arguments for the HTTP channel versus TCP or ZeroMQ is:
* It is possible to run multiple masters behind a single secure HTTPS load-balancer via hostname routing.
* It allows for reuse of the entire HTTP(S) ecosystem of tools, such as for throtttling, logging, tracing, etc.
* The protocol is widespread, and generally firewalls will accept outgoing HTTPS traffic, while random TCP trafic might be blocked.

Currently outstanding tasks:
* Configuration of the `http` channel, such as `master_uri`, potentially with a new format.
* Further refactoring of `salt/transport/abstract,py`.
* Minion connection / reconnection handling events within the `http` channel.
* Tests

### What issues does this PR fix or reference?
Relates to: #57088 

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [ ] Docs
- [ ] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [ ] Tests written/updated

### Commits signed with GPG?
Yes/No

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

**Comment by Skeen** [comment]
Should an RFC be made before this has any chance of being merged?

**Comment by lukasraska** [comment]
Great initiative, my 2 cents here (note that I haven't tested the implementation, just took a look at the source):
- long polling seems like generally a bad idea, there might be multiple connections at the same time, there can be time period between one request ends and second is established - resulting in missed messages, websockets would be better here in my option
- the general implementation very closely mimics what ZMQ transport does - publishes events to all connected minions, exactly what TCP transport avoids (see `publish_payload`) - while this does make sense for ZMQ as it's designed for that, HTTP transport could handle that better
- If I'm not mistaken, the payload of the HTTP message is URL-encoded JSON with base64-encoded msgpack data as `payload` field - not that it would necessarily be a bad idea (that way it reuses a LOT of existing code), but I don't understand the reasoning behind the implementation itself - the content cannot be easily validated / checked (for example by IDS/IPS systems)... essentially it's just TCP transport wrapped in HTTP session without the benefits of TCP transport itself - using something like JWT would be much better, but also much harder to implement

Overall I think it's a good idea to implement HTTP transport, but having it just a wrapper around existing logic doesn't help, it makes it more prone to errors. Having SEP for this would be probably good way to start, so Salt Core team can look into the proposal (I'm not affiliated with SaltStack in any way).

**Comment by Skeen** [comment]
HI, thanks you for your feedback!

> long polling seems like generally a bad idea, there might be multiple connections at the same time, there can be time period between one request ends and second is established - resulting in missed messages, ...

There shouldn't be missed messages, as the implementation uses a [`MessageBuffer`](https://github.com/saltstack/salt/pull/57210/files#diff-65f7ddbebf866e56c1903120ba9b530eR274-R298) to store messages as they come in, and as [clients request the server, with the id of the last message they've consumed](https://github.com/saltstack/salt/pull/57210/files#diff-65f7ddbebf866e56c1903120ba9b530eR250-R252), thus getting all messages that occured since their last connection (if any).
Because of this I do not think messages will be missed, unless the client checks in so infrequently, that the entire message buffer has been flushed through since their last connection. - Making the size of the message buffer configurable should solve this problem though.

> ... websockets would be better here in my option

I agree that websockets will be a good transport too, but I do not believe that one should exclude the other. As I see it, a polling HTTP transport can coexist with a websocket transport.

> the general implementation very closely mimics what ZMQ transport does - publishes events to all connected minions, exactly what TCP transport avoids (see `publish_payload`) - while this does make sense for ZMQ as it's designed for that, HTTP transport could handle that better

Agreed, and it is indeed very related to one of my outstanding tasks, namely 'minion connection / reconnection handling events within the `http` channel.'. The TCP tasks sends id's over to keep track of minions, while this implementation currently does not.

>If I'm not mistaken, the payload of the HTTP message is URL-encoded JSON with base64-encoded msgpack data as `payload` field - not that it would necessarily be a bad idea (that way it reuses a LOT of existing code), but I don't understand the reasoning behind the implementation itself - the content cannot be easily validated / checked (for example by IDS/IPS systems)... essentially it's just TCP transport wrapped in HTTP session without the benefits of TCP transport itself - using something like JWT would be much better, but also much harder to implement

You are not mistaken, what I went for was exactly the low-hanging fruit of just reusing as much existing code as possible. The reasoning behing this, was to get a fully functional implementation as soon as possible, as to test out routing and tracing using standard HTTP software.

The reason I created this PR early (and as a work-in-progress), is exactly to have these discussions and get this input, and it seems that here we've hit upon an interesting discussion indeed. Namely whether the transports should just do exactly what it says on the tin: Transport the data across the wire, or whether they should be specialized for their specific flavor, strengths and weaknesses.

I can see compelling arguments for both sides. Implementing the HTTP transport using JWT would utliize the strengths of HTTP better, however it would share less code and thus be harder to maintain, and would move a lot of the channel protocol into the transport instead of in an intermediate shared layer, as I suggest with a layer based architecture, where data flows through the layers towards the transport itself.

> Overall I think it's a good idea to implement HTTP transport, but having it just a wrapper around existing logic doesn't help, it makes it more prone to errors.

This is interesting, as i would argue for the opposite, the more code which is shared, the more well tested that code will be, and thus the less prone to errors, therefore minimizing the transport specific code and simultaneously minimizing the amount of transport specific bugs.

> Having SEP for this would be probably good way to start, so Salt Core team can look into the proposal (I'm not affiliated with SaltStack in any way).

Yea I thought it might be a good idea too.

File: salt/transport/http.py [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+    def __new__(cls, opts, **kwargs):
+        return super().__new__(cls, opts, **kwargs)
+
+    @classmethod
+    def __key(cls, opts, **kwargs):
+        return super().__key(cls, opts, **kwargs)
+
+    def start_channel(self, io_loop, **kwargs):
+        parse = urlparse.urlparse(self.opts["master_uri"])
+        master_host, master_port = parse.netloc.rsplit(":", 1)
```
Comment (frebib) [inline salt/transport/http.py:55]: This doesn't work for IPv6 addresses. Use `parse.hostname` and `parse.port` instead

File: salt/transport/http.py [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+                self.http_server.shutdown()
+            except Exception as exc:  # pylint: disable=broad-except
+                log.exception(
+                    "HTTPReqServerChannel close generated an exception: %s", str(exc)
+                )
+        elif hasattr(self.http_server, "stop"):
+            self.http_server.stop()
+
+    def _start_socket(self):
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Comment (frebib) [inline salt/transport/http.py:159]: No IPv6 love here

**Comment by waynew** [comment]
Is the goal here simply to add a new/abstract transport process, or are we explicitly concerned with security?

If it's the latter we should (MUST?) add tests for well-known attacks. Granted, that might be a /large/ suite of tests, but assuming they can be run in <24h that wouldn't be onerous to run periodically, and we would add it as a prerequisite for every release.

**Pull request closed without merging**